src/rdf4cpp/rdf/writer/BufWriter.hpp

```diff
@@ -43,27 +43,69 @@ struct Cursor {
     }
 };
 
-template<typename T>
-concept Serializer = requires (T &ser) {
-    typename T::Output;
-    typename T::Buffer;
-
-    { ser.buffer() } -> std::same_as<typename T::Buffer &>;
-    { ser.cursor() } -> std::same_as<Cursor &>;
-    { ser.finalize() } -> std::convertible_to<typename T::Output>;
-    T::flush(&ser.buffer(), ser.cursor());
+/**
+ * A type that can be used for buffered output.
+ * Conceptually, it is three things.
+ * Some buffer-like object a cursor into that buffer and a way to flush the buffer somewhere to make some room.
+ *
+ * Users of this concept write to the provided buffer in the place where the cursor points.
+ * After writing the cursor should be advanced.
+ * If the cursor says that there is no more room flush should be called.
+ */
+template<typename W>
+concept BufWriter = requires (W &bw) {
+    /**
+     * Return type of finalize, typically void
+     */
+    typename W::Output;
+
+    /**
+     * Type of the internal buffer
+     */
+    typename W::Buffer;
+
+    /**
+     * The underlying buffer, the only thing this need to be able to do is be used in flush
+     */
+    { bw.buffer() } -> std::same_as<typename W::Buffer &>;
+
+    /**
+     * The cursor pointing inside of w.buffer()
+     */
+    { bw.cursor() } -> std::same_as<Cursor &>;
```

Comment:
I am missing a function that allows to write with a BufWriter. 

it could look like:
```
pair<size_t, bool> write(void* in_buf, size_t len)
```

and return the number of bytes written and if the buffer is exhausted (flush is required). 
Alternatively, it could also take care of the flushes. 

I am not sure where to actually implement it. The best spot would probably be cursor with code similar to that one:
```
auto const max_write = std::min(str.size(), cursor.size());
        memcpy(cursor.data(), str.data(), max_write);
        cursor.advance(max_write);
```

(from write_str below)

Reply:
yes that sounds good.